content/blog/2021/06/writing-your-first-policy-with-kubewarden.md

```diff
@@ -0,0 +1,554 @@
+
+That's it for now. I am sure you already have in mind several ways to
+enrichen, improve and secure this simple policy.
+
+## The language
+
+At the time of writing we have SDKs for the following languages:
+
+- [Rust](https://github.com/kubewarden/policy-sdk-rust)
+- [Swift](https://github.com/kubewarden/policy-sdk-swift)
+- [TinyGo](https://github.com/kubewarden/policy-sdk-go)
+
+This list is not static and more will be implemented over time. If you
+want to know more about the SDKs and the languages we support and why,
+you can [extend the read through the Kubewarden
+book](https://docs.kubewarden.io/writing-policies/index.html#programming-language-requirements).
+
+We are going to write our policy in Go, thus, we will use the TinyGo
+compiler.
+
+In general, starting a new policy is a matter of using the policy
+template we have on GitHub, except for Rust, that has a dedicated
+tool. You can find more information [in the
+book](https://docs.kubewarden.io/).
+
+## The policy
+
+We head to the [`go-policy-template`
+repo](https://github.com/kubewarden/go-policy-template) and use that
+template, creating a repository in one of our GitHub organizations.
+
+### Prerequisites
+
+We clone the repository we just generated from the template locally.
+
+> **Note**: you will need `docker` or `podman` installed for this
+> exercise. At the time of writing, the policy template's `Makefile`
+> calls to the `docker` CLI, so if you use `podman`, make sure you
+> also have a `docker` wrapper or symlink in your `$PATH`. This is so
+> you don't have to install TinyGo yourself.
+
+Let's perform a sanity check by running `make`:
+
+```console
+$ rm -rf vendor
+$ make
+```
+
+> **Note**: we remove the `vendor` folder so dependencies will be
+> fetched on every build. This is just for the sake of ease of build
+> for the blog post: you don't need to do `go mod tidy` and `go mod
+> vendor` afterwards when we add more dependencies, so you don't need
+> to have any kind of Go toolchain with the right versions on your
+> machine -- a simple `make` will build it, a little bit slower than
+> vendored, but still. --
+
+We can see that as a result of the previous command, the sample policy
+that the template contains has been built in a file called
+`policy.wasm`.
+
+```console
+$ file policy.wasm
+policy.wasm: WebAssembly (wasm) binary module version 0x1 (MVP)
+```
+
+Looks about right! Now, let's go ahead and write our policy!
+
+### Policy first go
+
+Let's start by editing the `main.go` file. For the sake blog post
+brevity, I'll keep the code on the `main.go` file, so it's
+straightforward to follow. Also, I'll be skipping some other best
+practices, like TDD or even having unit tests, again for the sake of
+brevity.
+
+```go
+package main
+
+import (
+	wapc "github.com/wapc/wapc-guest-tinygo"
+	kubewarden "github.com/kubewarden/policy-sdk-go"
+)
+
+func main() {
+	wapc.RegisterFunctions(wapc.Functions{
+		"validate": func(payload []byte) ([]byte, error) {
```

Comment:
Add comments to the code, explaining why the line exists and what it does.

Reply:
The idea of the blog post is not to go too deep into the details. As many introductory blog posts, sometimes it takes you some code to copy and paste, and while you don't understand everything, you understand what you change and what you implement. Is a balance between understanding everything (that cannot be explained in an introductory post), and seeing results of your changes on the code. I fear that if I explain too much, this would become an introductory bible...

Reply:
I think we can find a middle ground. In this case I would do something like adding a single  comment inside of the `validate` and `validate_settings` blocks.

Something like:

```
// function validating the k8s request
```
...
```
// function validating policy settings
```